### DISCUSSION TITLE
UDAF adapter supporting companion functions of different aggregation steps

### DISCUSSION THREAD

**Discussion by kagamiori** [facebookincubator/velox#4493] [Ideas]
(With @mbasmanova)

We are designing a UDAF adapter to allow more flexible combination of different aggregation steps in aggregation queries. Potential usages include rolling aggregation and the support for Spark query plans for aggregation with one distinct (https://github.com/facebookincubator/velox/issues/4412). We discuss our work in progress in this discussion. 

# Use Cases
Some Velox users are interested in computing aggregation on a rolling basis, e.g., aggregate events happening within a 30-day window and update the aggregation result every day.

timestamp | event
-- | --
2023-01-03 04:00:00 | 15
2023-01-03 18:00:00 | 4
2023-01-04 11:00:00 | 20
2023-01-05 07:00:00 | 16
2023-01-05 20:00:00 | 7
… | …

A naive approach is to first filter records by date stamps and keep only those within the last 30 days and then aggregate events of them.
```
SELECT key, avg(event) 
FROM table 
WHERE date(timestamp) > '<DATEID-30>'
GROUP BY key;
```

However, computing such rolling aggregation every day from scratch is expensive and a waste of resources because the aggregation for events within the past few days won’t change as the window slides. 

Alternatively, one can pre-aggregate events every day to intermediate states and then aggregate these daily states together when computing a 30-day aggregation. In the example below, avg_partial computes intermediate states from input events, then  avg_merge aggregates partial states into another intermediate state, and finally, avg_extract converts the intermediate state into an aggregation result. 
```
INSERT INTO daily
SELECT key, ds, avg_partial(event) as state
FROM (SELECT key, date(timestamp) as ds, event FROM table)
GROUP BY key, ds;

SELECT key, avg_extract(avg_merge(state))
FROM daily
WHERE ds > '<DATEID-30>'
GROUP BY key;
```

This approach requires more flexible computation of different phases of an aggregation. Presto supports this by providing companion aggregate and scalar functions for a UDAF that allows aggregating raw inputs to intermediate states, combining intermediate states, and extracting the aggregation result from an intermediate state separately. But this is only supported for a small set of Presto UDAFs right now, such as approx_distinct() via [HyperLogLog functions](https://prestodb.io/docs/current/functions/hyperloglog.html) and approx_percentile() via [Q-Digest](https://prestodb.io/docs/current/functions/qdigest.html) and [T-Digest functions](https://prestodb.io/docs/current/functions/tdigest.html). 

This alternative approach is better, but it still incurs unnecessary computation when aggregating daily states from DATEID-29 to DATEID-1 because these do not change from the previous 30-day aggregation to the current one. An even better approach is to subtract the daily aggregation of the oldest day from the previous 30-day aggregation and add the newest daily aggregation to it.

<img width="996" alt="Screenshot 2023-03-28 at 2 23 10 PM" src="https://user-images.githubusercontent.com/16028358/229592851-d6b3e820-5e11-41cc-b6d8-5f480feb4746.png">

Below is an example query for rolling average over 30 days. In this example, window-size states are stored in the longterm table and slide-size states are stored in the daily table.
```
INSERT INTO daily
SELECT key, ds, avg_partial(event) 
FROM events
GROUP BY key, ds;

WITH u AS (
    SELECT key, state, FALSE AS to_add
    FROM daily
    WHERE ds = '<DATEID-30>'
    UNION ALL
    SELECT key, state, TRUE
    FROM longterm
    WHERE ds = '<DATEID-1>'
    UNION ALL
    SELECT key, state, TRUE
    FROM daily
    WHERE ds = '<DATEID>'
)
SELECT
    key,
    avg_extract(avg_retract(positive_merged, negative_merged))
FROM (
    SELECT
        key,
        avg_merge(state) FILTER (WHERE to_add = TRUE) AS positive_merged,
        avg_merge(state) FILTER (WHERE to_add = FALSE) AS negative_merged
    FROM u
    GROUP BY 1
)
```

# Goal
We believe this is a general use case that would benefit from a generic solution. Therefore, we propose to enhance Velox to automatically generate companion functions for all aggregate functions. Given a Velox UDAF, e.g., avg, Velox will automatically provide two companion scalar functions and two companion aggregate functions, together with their registration code.

Function Name | Type | Example | Description
-- | -- | -- | --
{AGGR_FUNC} | Aggregate | avg | The original UDAF.
{AGGR_FUNC}_partial | Aggregate | avg_partial | raw events -> intermediate state
{AGGR_FUNC}_merge | Aggregate | avg_merge | intermediate states -> intermediate state
{AGGR_FUNC}_extract | Scalar | avg_extract[^1] | intermediate state -> final result
{AGGR_FUNC}_retract | Scalar | avg_retract | (intermediate state a, intermediate state b) -> intermediate state (a - b)

[^1]: The extract function name may have a suffix of the result type name if the original UDAF has multiple signatures with the same intermediate type but different result types. E.g., avg() has two signatures real -> row(double, bigint) -> real and double -> row(double, bigint) -> double. Their extract function signatures are row(double, bigint) -> real and row(double, bigint) -> double. However, Velox requires an expression result type to be infer-able from the function name and its argument types. Therefore, we append a suffix of the result type name to the extract functions to differentiate them, i.e., avg_extract_real() and avg_extract_double().

# Design
This goal has two pieces: 
1. Auto-generate companion functions. 
2. Auto-generate registration code for them. 

For the first piece, Velox will extend the [vector-based UDAF authoring interface](https://facebookincubator.github.io/velox/develop/aggregate-functions.html) with two new optional APIs to allow defining how raw input and slide-size state can be retracted from a window-size state. UDAF authors will write UDAFs using this vector-based UDAF interface and Velox will provide an adapter to auto-generate companion functions for it.

For the second piece, the original registerAggregationFunction() API will be extended to accept an additional boolean flag indicating whether all companion functions will be registered together.

## Current Velox UDAF interface and registration
The current Velox UDAF authoring interface consists of seven core methods for aggregation that UDAF authors need to define. 

Methods | Descriptions
-- | --
initializeNewGroups() | How to create and initialize accumulators before aggregation starts.
addRawInput | How to add raw input data to accumulators for the corresponding groups.
addSingleGroupRawInput | Similar to addRawInput but assumes there is only one group.
extractAccumulators | How to extract a vector of intermediate states from accumulators so that intermediate states can optionally be shuffled to different nodes for subsequent aggregation steps.
addIntermediateResults | How to add intermediate states from an input vector to accumulators. The input vector may be shuffled from a different node by the previous aggregation step.
addSingleGroupIntermediateResults | Similar to addIntermediateResults but assumes there is only one group.
extractValues | How to extract a vector of final aggregation results from accumulators.

In addition to these methods, the UDAF author defines all supported signatures for this UDAF, including their input types, intermediate types, and the result types. These signatures are used to resolve data types at runtime. The author also defines a factory function that takes the input type, result type, and the current aggregation step (e.g., core::AggregationNode::Step::kPartial), and returns an instance of the UDAF that supports this use case.  

More details about how to implement a Velox UDAF can be found here: https://facebookincubator.github.io/velox/develop/aggregate-functions.html.

## Planned extensions
To support creation of the retract function, Velox will introduce two optional method functions to the UDAF authoring interface, namely retractRawInput and retractIntermediateResults. retractIntermediateResults will be used to generate the retract companion function, and retractRawInput will be used to optimize the computation of window functions. 
```
virtual void retractIntermediateResults(
     char** group,
     const SelectivityVector& rows,
     const std::vector<VectorPtr>& args) {
  VELOX_NYI();
}

virtual void retractRawInput(
     char** group,
     const SelectivityVector& rows,
     const std::vector<VectorPtr>& args) {
  VELOX_NYI();
}
```

The registration API will register all supported companion functions by default. The code that UDAF authors write is the same as before, but the partial, merge, and extract companion functions are registered automatically.

Not all UDAFs support the retract operation, e.g., min() and max(), hence we don’t register the retract companion function by default. If a UDAF supports retracting, the author should provide an AggregateFunctionMetadata object specifying this support when registering the UDAF. An extra retract function will then be registered together. 
```
struct AggregateFunctionMetadata {
 bool supportsRetract = false;
 bool isOrderSensitive = false;
};

/// Register an aggregate function with the specified name and signatures. If registerCompanionFunctions is true, also register companion aggregate and scalar functions with it.
bool registerAggregateFunction(
   const std::string& name,
   std::vector<std::shared_ptr<AggregateFunctionSignature>> signatures,
   AggregateFunctionFactory factory,
   AggregateFunctionMetadata metadata = {},
   bool registerCompanionFunctions = true);
```

An additional API will be provided for retrieving signatures of companion functions of a given UDAF. If a UDAF with the given name doesn’t exist, this API returns std::nullopt.
```
struct CompanionSignatureEntry {
 std::string functionName;
 std::vector<FunctionSignaturePtr> signatures;
};

enum class CompanionType : int8_t { kPartial, kMerge, kExtract, kRetract };

std::optional<
   std::unordered_map<CompanionType, std::vector<CompanionSignatureEntry>>>
getCompanionFunctionSignatures(const std::string& name);
```

## Known limitations
There are two known limitations with the current design. However, we expect most UDAFs to not have these issues or can avoid these issues through simple twists of their implementations.
1. If a signature of the original UDAF has type variables (e.g., T), its result type must be resolvable given solely the concrete intermediate type. E.g., T -> Varbinary -> T is not allowed, but T -> array(T) -> T is allowed.
2. If the original UDAF has multiple signatures with the same intermediate type but different result types, multiple extract functions will be generated, one for each distinct result type. The extract function names will be {AGGR_NAME}\_extract\_{RESULT_TYPE_NAME}. E.g., for avg(), there will be avg_extract_real() and avg_extract_double(). For a function in this situation that returns a complex type, e.g., Map<Varchar, Array<Bigint>>, we append the preorder traversal of the complex result type as the suffix to the extract function name, e.g., {AGGR_NAME}_extract_map_varchar_array_bigint.

## Generate companion functions
The table below shows how the companion functions are composed of the member functions from the original UDAF. Let’s assume the original UDAF is avg() that is implemented through a class AverageAggregate.

### Aggregate companion functions
Methods | avg_partial(arg) | avg_merge(arg)
-- | -- | --
initializeNewGroups | AverageAggregate::initializeNewGroups | AverageAggregate::initializeNewGroups
addRawInput | AverageAggregate::addRawInput | AverageAggregate::addIntermediateResults
addSingleGroupRawInput | AverageAggregate::addSingleGroupRawInput | AverageAggregate::addSingleGroupIntermediateResults
extractAccumulators | AverageAggregate::extractAccumulator | AverageAggregate::extractAccumulator
addIntermediateResults | AverageAggregate::addIntermediateResults | AverageAggregate::addIntermediateResults
addSingleGroupIntermediateResults | AverageAggregate::addSingleGroupIntermediateResults | AverageAggregate::addSingleGroupIntermediateResults
extractValues | AverageAggregate::extractAccumulator | AverageAggregate::extractAccumulator

### Scalar companion function
Functions | Implementations
-- | --
avg_retract(arg0, arg1) | char** groups = allocateGroups(allocator);<br><br>AverageAggregate::initializeNewGroups(groups, range);<br>AverageAggregate::addIntermediateResults(groups, rows, {args[0]});<br>AverageAggregate::retractIntermediateResults(groups, rows, {args[1]});<br>AverageAggregate::extractAccumulators(groups, rows.size(), args);<br><br>freeGroups(groups);
avg_extract(arg) | char** groups = allocateGroups(allocator);<br><br>AverageAggregate::initializeNewGroups(groups, range);<br>AverageAggregate::addIntermediateResults(groups, rows, args);<br>AverageAggregate::extractValues(groups, rows.size(), result);<br><br>freeGroups(groups);

To generate these UDFs and UDAFs, Velox will build an adapter that takes the original UDAF and creates all the companion functions. A prototype of the adapter can be found here: https://github.com/facebookincubator/velox/blob/fc41b12c27d6c9f06c145fbd938d80f1479174b7/velox/exec/AggregateFunctionAdapter.h#L27-L293

## Generate registration code
A UDAF author needs to write the registration code of the original UDAF and the rest registration of companion functions will be handled by a Velox adapter automatically. Specifically, the author needs to define all supported function signatures and a factory function that creates a std::unique_ptr of the original UDAF instance given the input type, result type, and the aggregation step ([example](https://github.com/facebookincubator/velox/blob/0e5889c0d2f048b6dee4e1b09590be1535cc43f7/velox/functions/prestosql/aggregates/AverageAggregate.cpp#L324-L427))

We extend the registerAggregateFunction() API so that companion functions are registered together with the original UDAF: https://github.com/facebookincubator/velox/blob/fc41b12c27d6c9f06c145fbd938d80f1479174b7/velox/exec/Aggregate.cpp#L55-L80 
The new API for retrieving companion function signatures is added together: https://github.com/facebookincubator/velox/blob/fc41b12c27d6c9f06c145fbd938d80f1479174b7/velox/exec/Aggregate.cpp#L106-L162

The RegisterAdapter generates factories that create an instance of the original UDAF and pass it to the companion function constructors. A prototype can be found here: https://github.com/facebookincubator/velox/blob/fc41b12c27d6c9f06c145fbd938d80f1479174b7/velox/exec/AggregateFunctionAdapter.h#L295-L480

Finally, https://github.com/facebookincubator/velox/pull/4489 gives an example of the aforementioned adapters together with unit tests for the auto-generated and auto-registered companion functions for avg().

**Comment by kagamiori** [comment]
cc @mbasmanova @rui-mo

**Comment by pedroerp** [comment]
Thank for putting together the detailed design plan. +1 for adding the feature.

**Comment by laithsakka** [comment]
thanks wei for the well written design for the point:
``
If the original UDAF has multiple signatures with the same intermediate type but different result types, multiple extract functions will be generated, one for each distinct result type. The extract function names will be {AGGR_NAME}extract{RESULT_TYPE_NAME}.
``` 
I wonder if {AGGR_NAME}extract{INPUT_TYPE(s)_NAME}.
makes more sense?
basically naming the functions by its input type instead of output.

**Reply by kagamiori** [reply]
I think both are doable for us, so it depends on the use cases. The extract function takes input of the intermediate type and returns the result type of the original UDAF. So if we add the original input type into its function name (which can be irrelevant to the input and output type of the extract function itself), it might not be straightforward to understand what that type is.

Also, as @mbasmanova pointed out offline, one UDAF may have multiple inputs, so having input types in the function names may make function names too long.

**Comment by laithsakka** [comment]
would it make sense to have average_merge_final. (an aggregation that starts with partial result and  up with final results with out having to call extract).

**Reply by kagamiori** [reply]
Yes, we can add this as an optimization after we have the basic functionalities to avoid the conversion such as accumulator --> intermediate result vector --> accumulator.

**Comment by laithsakka** [comment]
the document does not mention how addSingleGroupIntermediateResults is going to be created for the new agg function
@kagamiori

**Reply by kagamiori** [reply]
That would be similar to addIntermediateResults, but just call the corresponding "single group" APIs in the original UDAF. Let me add these to the table.

**Comment by laithsakka** [comment]
looks like we can also push    
const std::string& name,
std::vector<std::shared_ptr<AggregateFunctionSignature>> signatures,
into the meta data?

**Reply by kagamiori** [reply]
Could you elaborate a bit? Do you mean the name and signatures of the original UDAF or the companion functions? Also are you thinking about a new metadata abstraction or some existing ones?
